lib/Conversion/ImportVerilog/Structure.cpp

```diff
@@ -459,15 +470,77 @@ Context::convertCompilation(slang::ast::Compilation &compilation) {
 /// the op to the worklist of module bodies to be lowered. This acts like a
 /// module "declaration", allowing instances to already refer to a module even
 /// before its body has been lowered.
-ModuleLowering *
-Context::convertModuleHeader(const slang::ast::InstanceBodySymbol *module) {
+ModuleLowering *Context::convertModuleHeader(
+    const slang::ast::InstanceBodySymbol *module,
+    std::optional<slang::span<const slang::ast::PortConnection *const>>
+        portConnections) {
   using slang::ast::ArgumentDirection;
   using slang::ast::MultiPortSymbol;
+  using slang::ast::ParameterSymbol;
   using slang::ast::PortSymbol;
+  using slang::ast::TypeParameterSymbol;
+
+  auto parameters = module->parameters;
+  bool hasModuleSame = false;
+  // If there is already exist a module that has the same name with this
+  // module ,has the same parent scope and has the same parameters we can
+  // define this module is a duplicate module
+  for (auto const &existingModule : modules) {
+    if (module->getDeclaringDefinition() ==
+        existingModule.getFirst()->getDeclaringDefinition()) {
+      auto moduleParameters = existingModule.getFirst()->parameters;
+      hasModuleSame = true;
+      for (auto it1 = parameters.begin(), it2 = moduleParameters.begin();
+           it1 != parameters.end() && it2 != moduleParameters.end();
+           it1++, it2++) {
+        // Parameters size different
+        if (it1 == parameters.end() || it2 == moduleParameters.end()) {
+          hasModuleSame = false;
+          break;
+        }
+        const auto *para1 = (*it1)->symbol.as_if<ParameterSymbol>();
+        const auto *para2 = (*it2)->symbol.as_if<ParameterSymbol>();
+        // Parameters kind different
+        if ((para1 == nullptr) ^ (para2 == nullptr)) {
+          hasModuleSame = false;
+          break;
+        }
+        // Compare ParameterSymbol
+        if (para1 != nullptr) {
+          hasModuleSame = para1->getValue() == para2->getValue();
+        }
+        // Compare TypeParameterSymbol
+        if (para1 == nullptr) {
+          auto para1Type = convertType(
+              (*it1)->symbol.as<TypeParameterSymbol>().getTypeAlias());
+          auto para2Type = convertType(
+              (*it2)->symbol.as<TypeParameterSymbol>().getTypeAlias());
+          hasModuleSame = para1Type == para2Type;
+        }
+        if (!hasModuleSame)
+          break;
+      }
+      if (hasModuleSame) {
+        module = existingModule.first;
+        break;
+      }
+    }
+  }
 
   auto &slot = modules[module];
-  if (slot)
+  if (slot) {
+    if (portConnections) {
+      for (const auto *con : *portConnections) {
+        auto *port = con->port.as_if<PortSymbol>();
+        for (const auto &existingPort : slot->ports)
+          if (port->name == existingPort.ast.name)
+            slot->portsBySyntaxNode.insert(
+                {port->getSyntax(), &existingPort.ast});
+      }
+    }
```

Comment:
I don't think you need the `portConnections` for this. The `portsBySyntaxNode` should contain a mapping from the syntax node of a port to the port of the module. Something like:
```suggestion
    for (const auto &port : slot->ports)
      slot->portsBySyntaxNode.insert({port.ast.getSyntax(), &port.ast});
```
When you're dealing with instances, you can use the `portsBySyntaxNode` map to map from the instance's duplicated port to the original port on the already-lowered module. Something like:
```
auto *modulePort = moduleLowering->portsBySyntaxNode.lookup(instancePort->getSyntax());
```
The instancePort is a copy of modulePort. They are separate objects, but I think their syntax is a pointer to the exact same AST node. That's what we use as a key to map from the instance's duplicated port to the original port of the lowering.

Reply:
That's a really deep understanding which I can not get this information from AST.json! I was thinking to mapping duplicate instance port symbol to existing instance's relating module port symbol, which is somehow complex. I have imporve my code.